Sts1CobcSw/Edu/ProgramStatusHistory.cpp

```diff
@@ -1,24 +1,29 @@
 #include <Sts1CobcSw/Edu/ProgramStatusHistory.hpp>
+#include <Sts1CobcSw/FramSections/FramLayout.hpp>
+#include <Sts1CobcSw/FramSections/RingArray.hpp>
 
 #include <strong_type/equality.hpp>
 
 
 namespace sts1cobcsw::edu
 {
-RODOS::RingBuffer<ProgramStatusHistoryEntry, nProgramStatusHistoryEntries> programStatusHistory;
+constexpr auto programStatusHistorySection = framSections.Get<"eduProgramStatusHistory">();
+sts1cobcsw::RingArray<ProgramStatusHistoryEntry,
+                      programStatusHistorySection,
+                      nCachedProgramStatusHistoryEntries>
+    programStatusHistory;
 
 
 auto UpdateProgramStatusHistory(ProgramId programId, RealTime startTime, ProgramStatus newStatus)
     -> void
 {
-    // TODO: Check that there is only one entry matching program/queue ID, or should it be the case
-    // by construction ?
-    for(std::uint32_t i = 0; i < programStatusHistory.occupiedCnt; ++i)
+    for(std::size_t i = 0; i < programStatusHistory.Size(); ++i)
     {
-        if(programStatusHistory.vals[i].startTime == startTime
-           and programStatusHistory.vals[i].programId == programId)
+        auto entry = programStatusHistory.Get(i);
+        if(entry.startTime == startTime and entry.programId == programId)
         {
-            programStatusHistory.vals[i].status = newStatus;
+            entry.status = newStatus;
+            programStatusHistory.Set(i, entry);
```

Comment:
I just realized that this is not thread-safe in the sense that someone might call `PushBack()` between `Get()` and `Set()`. This would mean that the index `i` might point to a different location/entry. We somehow need to ensure that getting and setting the entry is atomic. The only way that I can think of right now that would achieve that is to add `RingArray<T>::FindAndReplace(predicate, newData/Element/Entry)` where `predicate` is a callable that takes a `T const &` and returns something convertible to `bool`.